src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs

```diff
@@ -298,50 +299,107 @@ async Task OnRunOrchestratorAsync(
-                        _ => null,
-                    };
+                        case DurableTaskWorkerOptions.VersionMatchStrategy.None:
+                            // No versioning, breakout.
+                            break;
+                        case DurableTaskWorkerOptions.VersionMatchStrategy.Strict:
+                            // Comparison of 0 indicates equality.
+                            if (versionComparison != 0)
+                            {
+                                failureDetails = new P.TaskFailureDetails
+                                {
+                                    ErrorType = "VersionMismatch",
+                                    ErrorMessage = $"The orchestration version '{runtimeState.Version}' does not match the worker version '{versioning.Version}'.",
+                                    IsNonRetriable = true,
+                                };
+                            }
+
+                            break;
+                        case DurableTaskWorkerOptions.VersionMatchStrategy.CurrentOrOlder:
+                            // Comparison > 0 indicates the orchestration version is greater than the worker version.
+                            if (versionComparison > 0)
+                            {
+                                failureDetails = new P.TaskFailureDetails
+                                {
+                                    ErrorType = "VersionMismatch",
+                                    ErrorMessage = $"The orchestration version '{runtimeState.Version}' is greater than the worker version '{versioning.Version}'.",
+                                    IsNonRetriable = true,
+                                };
+                            }
+
+                            break;
+                        default:
+                            // If there is a type of versioning we don't understand, it is better to treat it as a versioning failure.
+                            failureDetails = new P.TaskFailureDetails
+                            {
+                                ErrorType = "VersionError",
+                                ErrorMessage = $"The version match strategy '{versioning.MatchStrategy}' is unknown.",
+                                IsNonRetriable = true,
+                            };
+                            break;
+                    }
 
-                    TaskOrchestration shim = this.shimFactory.CreateOrchestration(name, orchestrator, parent);
-                    TaskOrchestrationExecutor executor = new(
-                        runtimeState,
-                        shim,
-                        BehaviorOnContinueAsNew.Carryover,
-                        request.EntityParameters.ToCore(),
-                        ErrorPropagationMode.UseFailureDetails);
-                    result = executor.Execute();
+                    versionFailure = failureDetails != null;
                 }
```

Comment:
Non-blocking: Would it be easy to move this version checking logic into a helper method? This method is getting kind of long and harder to reason about the end-to-end flow. Breaking this logic out into a helper method would make the outer `OnRunOrchestratorAsync` method a little easier for humans to parse.

Reply:
That's fair, moved the versioning check out.